shoop/admin/modules/products/views/forms.py

```diff
@@ -134,3 +142,115 @@ def save(self):
+
+        self.file_url = self.instance.url
+        self.thumbnail = None
+
+    def pre_master_save(self, instance):
+        instance.product = self.product
+
+
+class BaseProductMediaFormSet(BaseModelFormSet):
+    validate_min = False
+    min_num = DEFAULT_MIN_NUM
+    validate_max = False
+    max_num = DEFAULT_MAX_NUM
+    absolute_max = DEFAULT_MAX_NUM
+    model = ProductMedia
+    can_delete = True
+    can_order = False
+    extra = 1
+
+    def __init__(self, *args, **kwargs):
+        self.product = kwargs.pop("product")
+        self.languages = to_language_codes(kwargs.pop("languages", ()))
+        kwargs.pop("empty_permitted")  # this is unknown to formset
+        super(BaseProductMediaFormSet, self).__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        qs = ProductMedia.objects.filter(product=self.product)
+        if self.allow_only:
+            qs = qs.filter(kind=self.allow_only)
+        else:
+            qs = qs.exclude(kind=ProductMediaKind.IMAGE)
+        return qs
+
+    def form(self, **kwargs):
+        kwargs.setdefault("languages", self.languages)
+        kwargs.setdefault("product", self.product)
+        kwargs.setdefault("allow_only", self.allow_only)
+        return self.form_class(**kwargs)
+
+
+class ProductMediaFormSet(BaseProductMediaFormSet):
+    allow_only = None
+    form_class = ProductMediaForm
+
+
+class ProductImageMediaForm(ProductMediaForm):
+    def __init__(self, **kwargs):
+        super(ProductImageMediaForm, self).__init__(**kwargs)
+
+        self.fields["kind"].initial = self.allow_only
+        self.fields["kind"].widget = forms.HiddenInput()
+
+        if self.instance.pk and isinstance(self.instance, Image):
+            thumbnail = self.instance.easy_thumbnails_thumbnailer.get_thumbnail({
+                'size': (64, 64),
+                'crop': True,
+                'upscale': True,
+            })
+            self.file_url = self.instance.url
+            self.thumbnail = thumbnail.url or None
+
+    def clean_file(self):
+        file = self.cleaned_data.get("file")
+        if file and self.allow_only == ProductMediaKind.IMAGE and not isinstance(file, Image):
+            raise ValidationError(_("Only images allowed in this field"))
+        return file
+
+    def save(self, commit=True):
+        instance = super(ProductImageMediaForm, self).save(commit)
+        print(self.cleaned_data)
+        if self.cleaned_data.get("is_primary"):
+            self.product.primary_image = instance
+            self.product.save()
+        return instance
+
+
+class ProductImageMediaFormSet(ProductMediaFormSet):
+    allow_only = ProductMediaKind.IMAGE
+    form_class = ProductImageMediaForm
+
+    def add_fields(self, form, index):
+        super(ProductImageMediaFormSet, self).add_fields(form, index)
+        form.fields["is_primary"] = forms.BooleanField(required=False)
+        if form.instance and self.product.primary_image_id == form.instance.pk:
```

Comment:
Likewise this looks like something that should be done in the form's init instead of hijacking `add_fields`.